src/relentless/model/potential/angle.py

```diff
@@ -0,0 +1,306 @@
+        k = params["k"]
+        theta0 = params["theta0"]
+
+        theta, f, s = self._zeros(theta)
+
+        f = k * (numpy.cos(theta) - numpy.cos(theta0)) * numpy.sin(theta)
+
+        if s:
+            f = f.item()
+        return f
+
+    def _derivative(self, param, theta, k, theta0):
+        r"""Evaluate angle derivative with respect to a variable."""
+        theta, d, s = self._zeros(theta)
+
+        if param == "k":
+            d = 0.5 * (numpy.cos(theta) - numpy.cos(theta0)) ** 2
+        elif param == "theta0":
+            d = k * (numpy.cos(theta) - numpy.cos(theta0)) * numpy.sin(theta0)
+        else:
+            raise ValueError("Unknown parameter")
+
+        if s:
+            d = d.item()
+        return d
+
+
+class CosineAngle(AnglePotential):
+    r"""Cosine angle potential.
+
+    .. math::
+
+        u(\theta) = k (1 + cos(\theta))
+
+    where :math:`\theta` is the angle between three bonded particles. The parameters
+    for each type are:
+
+    +-------------+--------------------------------------------------+
+    | Parameter   | Description                                      |
+    +=============+==================================================+
+    | ``k``       | Spring constant :math:`k`.                       |
+    +-------------+--------------------------------------------------+
+
+    Parameters
+    ----------
+    types : tuple[str]
+        Types.
+    name : str
+        Unique name of the potential. Defaults to ``__u[id]``, where ``id`` is the
+        unique integer ID of the potential.
+
+    Attributes
+    ----------
+    coeff : :class:`AngleParameters`
+        Parameters of the potential for each type.
+
+    Examples
+    --------
+    Cosine Angle::
+
+        >>> u = relentless.potential.angle.CosineAngle(("A",))
+        >>> u.coeff["A"].update({'k': 1000})
+
+    """
+
+    def __init__(self, types, name=None):
+        super().__init__(keys=types, params=("k",), name=name)
+
+    def energy(self, type_, theta):
+        """Evaluate angle energy."""
+        params = self.coeff.evaluate(type_)
+        k = params["k"]
+
+        theta, u, s = self._zeros(theta)
+
+        u = k * (1 + numpy.cos(theta))
+
+        if s:
+            u = u.item()
+        return u
+
+    def force(self, type_, theta):
+        """Evaluate angle force."""
+        params = self.coeff.evaluate(type_)
+        k = params["k"]
+
+        theta, f, s = self._zeros(theta)
+
+        f = k * numpy.sin(theta)
+
+        if s:
+            f = f.item()
+        return f
+
+    def _derivative(self, param, theta, k):
+        r"""Evaluate angle derivative with respect to a variable."""
+        theta, d, s = self._zeros(theta)
+
+        if param == "k":
+            d = 1 + numpy.cos(theta)
+        else:
+            raise ValueError("Unknown parameter")
+
+        if s:
+            d = d.item()
+        return d
+
+
+class AngleSpline(potential.BondedSpline, AnglePotential):
```

Comment:
Following with the other comment: here we want to give the class a full docstring describing its constructor and examples. The description of the constructor arguments can be basically copy–pasted from the base class, then add specific examples.

I think you then don't really need the methods below, and can just give an empty class body with `pass` because all of these are just calling the parent methods. The class will inherit the docstrings of those parents, which are probably more detailed.

Same comment applies to the `BondSpline`.

Reply:
The methods below change the `r` input to `theta`. Thinking more on this the rs are hardcoded in as parameters as "dr-", so this doesn't eliminate all of them. 

If we just pass, are we fine having just using r in all splines?

Reply:
Oh good point. I think in the base potential tabulator, we use x as a generic coordinate so you could follow the same practice here. I’ll leave it up to you whether you want to just use x for all derived classes, or rename the arguments like you’ve done here.

If you rename the arguments, you can also rename the string name of those parameters by adding a class variable like `_space_coord_name = “x”` in the base class , then override it in the deriving classes to r, theta, etc. You would use this variable to format the parameter names as dr, dtheta, etc.